crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+
+        let projection = Projection::builder()
+            .projection_type(ProjectionType::All)
+            .build();
+
+        let gsi = GlobalSecondaryIndex::builder()
+            .index_name(STATUS_RELAY_INDEX)
+            .key_schema(gsi_hash_key)
+            .key_schema(gsi_range_key)
+            .projection(projection)
+            .provisioned_throughput(gsi_throughput)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("GSI build error: {e}")))?;
+
+        if let Err(e) = self
+            .client
+            .create_table()
+            .table_name(TABLE_NAME)
+            .attribute_definitions(attr_def_id)
+            .attribute_definitions(attr_def_status)
+            .attribute_definitions(attr_def_relay_after)
+            .key_schema(key_schema)
+            .global_secondary_indexes(gsi)
+            .provisioned_throughput(throughput)
+            .send()
+            .await
+        {
+            return Err(StorageError::Internal(format!(
+                "Failed to create DynamoDB table: {e}"
+            )));
+        }
+
+        Ok(())
+    }
+
+    async fn query_requests_by_status_and_time(
+        &self,
+        status: RequestStatus,
+        max_timestamp: &str,
+        limit: usize,
+    ) -> Result<Vec<ScheduledRequest>, StorageError> {
+        let mut requests = Vec::new();
+        let mut last_evaluated_key = None;
+
+        loop {
+            let mut query_builder = self
+                .client
+                .query()
+                .index_name(STATUS_RELAY_INDEX)
+                .key_condition_expression("#status = :status AND relay_after <= :max_time")
+                .expression_attribute_names("#status", "status")
+                .expression_attribute_values(
+                    ":status",
+                    AttributeValue::S(status_to_str(&status).to_string()),
+                )
+                .expression_attribute_values(
+                    ":max_time",
+                    AttributeValue::N(max_timestamp.to_string()),
+                )
+                .limit((limit - requests.len()) as i32);
+
+            if let Some(key) = last_evaluated_key {
+                for (attr_name, attr_value) in key {
+                    query_builder = query_builder.exclusive_start_key(attr_name, attr_value);
+                }
+            }
+
+            let result = query_builder
+                .send()
+                .await
+                .map_err(|e| map_internal("query", e))?;
+
+            // Process items from this page
+            for item in result.items() {
+                if let Some(AttributeValue::S(payload_str)) = item.get("payload") {
+                    if let Ok(req) = serde_json::from_str::<ScheduledRequest>(payload_str) {
+                        requests.push(req);
+                        if requests.len() >= limit {
+                            return Ok(requests);
```

Comment:
Function signature lays to caller here :P I'd return only exactly `limit` of requests here. Dropping requests here is not a problem, as they would be eventually picked up by another scheduler iteration